test/test_policy.py

```diff
@@ -0,0 +1,812 @@
+            p.set_callback(policy_cb_types.CALC_NVPUBLIC, nvpublic_cb)
+            p.set_callback(policy_cb_types.EXEC_AUTH, bad_cb)
+            p.execute(self.ectx, session)
+        self.assertEqual(str(e.exception), "callback exception")
+
+    def test_exec_polsel_callback(self):
+        pol = {
+            "description": "this is a polsel policy",
+            "policy": [
+                {
+                    "type": "or",
+                    "branches": [
+                        {
+                            "name": "branch1",
+                            "description": "branch1 description",
+                            "policy": [{"type": "password"}],
+                        },
+                        {
+                            "name": "branch2",
+                            "description": "branch2 description",
+                            "policy": [{"type": "locality", "locality": ["zero",]}],
+                        },
+                    ],
+                },
+            ],
+        }
+        polstr = json.dumps(pol).encode()
+        session = self.ectx.start_auth_session(
+            ESYS_TR.NONE,
+            ESYS_TR.NONE,
+            TPM2_SE.POLICY,
+            TPMT_SYM_DEF(algorithm=TPM2_ALG.NULL),
+            TPM2_ALG.SHA256,
+        )
+        cb_auth_object = ""
+        cb_branches = None
+
+        def polsel_cb(auth_object, branches):
+            nonlocal cb_auth_object
+            nonlocal cb_branches
+            cb_auth_object = auth_object
+            cb_branches = branches
+            return len(branches) - 1
+
+        with policy(polstr, TPM2_ALG.SHA256) as p:
+            p.set_callback(policy_cb_types.EXEC_POLSEL, polsel_cb)
+            p.execute(self.ectx, session)
+
+        dig2b = self.ectx.policy_get_digest(session)
+        self.assertEqual(cb_auth_object, None)
+        self.assertEqual(cb_branches, [b"branch1", b"branch2"])
+        self.assertEqual(
+            dig2b,
+            b"8\x17\xfa\x84\x98\xf9\xf6\xa0s\xaa\xc5\x91r\x0b\xc4\xea\xdf3\xd6\xdb#\xd5\n\x05\x12\xd7\x8a\x84\xb5\xa3\xb2\xa1",
+        )
+
+        def bad_cb(*args):
+            raise Exception("callback exception")
+
+        with self.assertRaises(Exception) as e, policy(polstr, TPM2_ALG.SHA256) as p:
+            p.set_callback(policy_cb_types.EXEC_POLSEL, bad_cb)
+            p.execute(self.ectx, session)
+        self.assertEqual(str(e.exception), "callback exception")
+
+    def test_exec_sign_callback(self):
+        private_key = b"-----BEGIN PRIVATE KEY-----\nMG8CAQAwEwYHKoZIzj0CAQYIKoZIzj0DAQEEVTBTAgEBBBgBJYQqvoPfXctJixFL\nlAzRLQaAFBHOoQyhNAMyAATAjqP6LEx2q1p5aUSAfSwIpr0NijnvyLfWtluYrqCJ\nsI7HNirP/FKiz8pIY3FAD18=\n-----END PRIVATE KEY-----\n"
+        public_key = b"-----BEGIN PUBLIC KEY-----\nMEkwEwYHKoZIzj0CAQYIKoZIzj0DAQEDMgAEwI6j+ixMdqtaeWlEgH0sCKa9DYo5\n78i31rZbmK6gibCOxzYqz/xSos/KSGNxQA9f\n-----END PUBLIC KEY-----\n"
```

Comment:
This got missed, I left it as a regular comment instead of a code review comment, my mistake. Can you use multi line string literals ''' here and textwrap.dedent to make the formatting more tolerable.

Reply:
Sorry, I missed one and just fixed the other one